utils/watermark_disk.go

```diff
@@ -0,0 +1,114 @@
+package utils
+
+import (
+	"crypto/rand"
+	"fmt"
+	"io"
+	"math/big"
+	"os"
+	"slices"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	bufferSize    = 512
+	numWatermarks = 10
+)
+
+type watermark struct {
+	position int64
+	data     []byte
+}
+
+func ApplyWatermarks(logicalName string) func() error {
+	// Write open
+	file, err := os.OpenFile(logicalName, os.O_WRONLY, 0)
+	if err != nil {
+		return func() error {
+			return err
+		}
+	}
+	defer file.Close()
+
+	// Get disk or partition size
+	fileSize, err := file.Seek(0, io.SeekEnd)
+	if err != nil {
+		return func() error {
+			return err
+		}
+	}
+	// Write watermarks on random locations
+	watermarks := writeWatermarks(file, 0, fileSize, numWatermarks)
+	if len(watermarks) != numWatermarks {
+		return func() error {
+			ErrorWritingWatermarks := errors.New("Error writing watermarks in the file")
+			return fmt.Errorf("%s | %w", logicalName, ErrorWritingWatermarks)
+		}
+	}
+
+	checker := func() error {
+		file, err := os.OpenFile(logicalName, os.O_RDONLY, 0)
+		if err != nil {
+			return err
+		}
+		defer file.Close()
+
+		for _, watermark := range watermarks {
+			_, err = file.Seek(watermark.position, io.SeekStart)
+			if err != nil {
+				return err
+			}
+			// Read the watermark written to the position
+			currentValue := make([]byte, bufferSize)
+			_, err = io.ReadFull(file, currentValue)
+			if err != nil {
+				return err
+			}
+			// Check if the watermark is still in the disk
+			if slices.Equal(currentValue, watermark.data) {
+				ErrorExistingWatermark := errors.New("Error existing watermark in the file")
+				return fmt.Errorf("%s | %w", logicalName, ErrorExistingWatermark)
+			}
+		}
+		return nil
+	}
+	return checker
+}
+
+func writeWatermarks(file *os.File, a, b int64, count int) []watermark {
+	if count == 1 {
+		data := make([]byte, bufferSize)
+		_, err := rand.Read(data)
+		if err != nil {
+			return nil
+		}
+		offset, err := rand.Int(rand.Reader, big.NewInt(b-a-bufferSize))
+		if err != nil {
+			return nil
+		}
+		randomPosition := int64(offset.Uint64()) + a
+		_, err = file.Seek(randomPosition, io.SeekStart)
+		if err != nil {
+			return nil
+		}
+		_, err = file.Write(data)
+		if err != nil {
+			return nil
+		}
+
+		w := watermark{
+			position: randomPosition,
+			data:     data,
+		}
+		return []watermark{w}
+	}
+	// Divide the intervals into two equal parts (approximately)
+	mid := (a + b) / 2
+
+	// Return recursively the call of function with the two remaining intervals
+	leftCount := count / 2
+	rightCount := count - leftCount
+
+	return append(writeWatermarks(file, a, mid-1, leftCount), writeWatermarks(file, mid, b, rightCount)...)
```

Comment:
why recurse instead of loop?

Reply:
It came to my mind as a divide and conquer algorithm, it breaks the problem info into smaller ones, and you only need to implement the base case and split otherwise

https://upload.wikimedia.org/wikipedia/commons/e/e6/Merge_sort_algorithm_diagram.svg

Reply:
seems unnecessary and confusing to me. loop > recursion for readability most times imo :D. The function signature also shows that, what is a and whit is b?

Reply:
in math you define a closed interval with [a,b] in our case [0, fileSize]. if watermarks > 1 you split the interval in two [0, mid -1] and [mid, fileSize]. each interval is now smaller and the watermark count decreases for each call. I feel that using recursion here is an elegant and concise way to solve it. I'm not agree with loop > recursion for example fibonacci sequence

Reply:
Changed to iterative! Thx